tests/integration/src/test/java/org/apache/pulsar/tests/integration/functions/java/PulsarWorkerRebalanceDrainTest.java

```diff
@@ -0,0 +1,393 @@
+        //       + "pulsar-functions-worker-process-qmcao-0    "
+        //       + "[public/default/testDrainFunctionality-nsnqbnlc:0, public/default/testDrainFunctionality-tbakebis:0] ";
+
+        final int nextWorkerStart = 0;
+        String remainingJson = json;
+        String nextFunctionList;
+        String nextWorker;
+        boolean moreToParse = true;
+        List<Map<String, Collection<String>>> retVal = new ArrayList<>();
+
+        while (moreToParse) {
+            int nextFunctionListStart = remainingJson.indexOf("[");
+            int nextFunctionListEnd = remainingJson.indexOf("]");
+            nextWorker = remainingJson.substring(nextWorkerStart, nextFunctionListStart);
+            nextWorker = nextWorker.replaceAll("\\s+","");
+            nextFunctionList = remainingJson.substring(nextFunctionListStart + 1, nextFunctionListEnd);
+            String[] funcAssignments = nextFunctionList.split(",");
+            Map<String, Collection<String>> curMap = new HashMap<>();
+            curMap.put(nextWorker, Arrays.asList(funcAssignments));
+            log.info("Found new entry: {}, {}", nextWorker, funcAssignments);
+            retVal.add(curMap);
+            log.info("retVal is {}", retVal);
+
+            // If we are already at the end, or if there are no more opening brackets
+            // (corresponding to function assignments), we are done.
+            try {
+                remainingJson = remainingJson.substring(nextFunctionListEnd + 1);
+            } catch (Throwable t) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Got exception {} while moving past function-list-end", t.getMessage());
+                }
+                moreToParse = false;
+            }
+            if (remainingJson.indexOf("[") < 0) {
+                moreToParse = false;
+            }
+        }
+        return retVal;
+    }
+
+    private List<WorkerInfo> getClusterStatus() throws Exception {
+        val result = pulsarCluster.getAnyWorker().execCmd(
+                PulsarCluster.ADMIN_SCRIPT,
+                "functions-worker",
+                "get-cluster"
+        );
+        log.debug("getClusterStatus result is: {}", result);
+        return workerInfoDecode(result.getStdout());
+    }
+
+    private WorkerInfo getClusterLeader() throws Exception {
+        ContainerExecResult result = pulsarCluster.getAnyWorker().execCmd(
+                PulsarCluster.ADMIN_SCRIPT,
+                "functions-worker",
+                "get-cluster-leader"
+        );
+        List<WorkerInfo> winfos = workerInfoDecode(result.getStdout());
+        assertEquals(winfos.size(), 1);
+        return workerInfoDecode(result.getStdout()).get(0);
+    }
+
+    private List<Map<String, Collection<String>>> getFunctionAssignments() throws Exception
+    {
+        ContainerExecResult result = pulsarCluster.getAnyWorker().execCmd(
```

Comment:
The output of this puslar-admin cli call is suppose to be more human readable but not easy to parse.  If you call the REST API, the response will be in json and easier to parse.  Not a super big deal, just the parsing code will be much simpler.